File: 2023_07_31_functional_dependency_injection/fsharp_impl/FunctionalDependencyInjection/FunctionalDependencyInjection/Domain.fs
module Domain

open System

type Id = string

type PatientRegistrationInfoDto =
    | Anonymous of email: string
    | AnonymousWithCreatingNewAccount of email: string
    | LoggedIn of patientId: Id
    | LoggedInWithUpdatingEmail of patientId: Id * email: string

type RegisterVisitRequestDto = { SlotId: Id; Patient: PatientRegistrationInfoDto; ServiceId: Id }

type RegisterVisitResponseDto = { VisitId: Id; ExternalPaymentId: option<Id> } //Option<Id>

type Slot = { Id: string; From: DateTime; To: DateTime; ServiceIds: Id []; UnitId: Id; Price: decimal }

type Visit =
    { Id: Id
      SlotId: Id
      From: DateTime
      To: DateTime
      ServiceId: Id
      UnitId: Id
      PatientId: Id option
      PatientEmail: string }

type Patient = { Id: Id; Name: string; Email: string }

type Payment = { Id: Id; UserEmail: string; Price: decimal; VisitId: Id; ExternalPaymentId: string }

type RegistrationError =
    | SlotNotFound
    | SlotServiceNotFound
    | PatientNotFound

File: 2023_07_31_functional_dependency_injection/typescript_impl/src/domain.ts
import { Option } from "powerfp";

export type Id = string;

export type PatientRegistrationInfoDto =
    | { type: "Anonymous"; email: string; }
    | { type: "AnonymousWithCreatingNewAccount"; email: string; }
    | { type: "LoggedIn"; patientId: Id; }
    | { type: "LoggedInWithUpdatingEmail"; patientId: Id; email: string; }

export type RegisterVisitRequestDto = { slotId: Id; patient: PatientRegistrationInfoDto; serviceId: Id; }

export type RegisterVisitResponseDto = { visitId: Id; externalPaymentId: Option<Id>; }

export type Slot = { id: string; from: Date; to: Date; serviceIds: Id[]; unitId: Id; price: number; }

export type Visit = { id: Id; slotId: Id; from: Date; to: Date; serviceId: Id; unitId: Id; patientId: Option<Id>; patientEmail: string; }

export type Patient = { id: Id; name: string; email: string; }

export type Payment = { id: Id; userEmail: string; price: number; visitId: Id; externalPaymentId: string; }

export type RegistrationError = "SlotNotFound" | "SlotServiceNotFound" | "PatientNotFound"
